docs/source/common_issues.rst

```diff
@@ -821,3 +821,32 @@ This is best understood via an example:
 To get this code to type check, you could assign ``y = x`` after ``x`` has been
 narrowed, and use ``y`` in the inner function, or add an assert in the inner
 function.
+
+
+Upgrading to PEP584 syntax
+---------------------------
+
+In pre-PEP584 syntax type annotations for builtin types such as ``dict``, ``list``, ``tuple``, ``type`` were taken from the ``typing`` module as ``typing.Dict``, ``typing.List``, ``typing.Tuple``, ``typing.Type`` etc.
+
+When upgrading a codebase you may hit a case such as the following:
+
+.. code-block:: python
+
+    from typing import TypeVar
+
+    T = TypeVar('T', bound='Foo')
+
+    class Foo:
+	type: str
+
+	@classmethod
+	def foo(cls: type[T], value: str) -> None:  # will fail due to "not a valid type"
+	    pass
+
+This is happening since the class has an attribute that is overshadowing the built-in keyword ``type``. To get this code to type-check you have the following options:
```

Comment:
- `type` is a builtin name, but it's not a _keyword_ according to the language specification. (I _think_ this is the correct intersphinx reference to link to the CPython docs for `type`; apologies if not.)
- I think a slightly longer example in the interactive REPL might be helpful here to illustrate exactly what's going on with the scoping semantics.

```suggestion
This occurs due to the fact that the class has an attribute named `type`
that shadows the built-in function :py:func:`type`.
To illustrate: at runtime in the following example,
the annotation for `arg` in the `ham` method 
is resolved as being the string `"spam"`
due to the fact that method signatures are executed
in the same scope as class attributes:

.. code-block:: pycon

    >>> class Eggs:
    ...     type: str = "spam"
    ...     def ham(self, arg: type) -> None:
    ...         pass
    ...
    >>> import inspect
    >>> inspect.signature(Eggs.ham)
    <Signature (self, arg: 'spam') -> None>

Mypy follows the same scoping semantics as Python at runtime
when resolving references.
To get this code to type-check, therefore, you have the following options:
```

Reply:
> type is a builtin name, but it's not a keyword according to the language specification. (I think this is the correct intersphinx reference to link to the CPython docs for type; apologies if not.)

True. Thanks will change it. The issue I have with the link is that it does not work/render when creating the html docs. I tried to grep mypy docs for usage of type and I already see the same problem at https://mypy.readthedocs.io/en/stable/type_narrowing.html. The `type()` link is the only one not working. It's using the same exact schema you gave here.

>  think a slightly longer example in the interactive REPL might be helpful here to illustrate exactly what's going on with the scoping semantics.

Very nice idea. I agree and done it. It also helped me better understand it :)

One question, what does `pycon` stand for in this code block?

Reply:
> One question, what does `pycon` stand for in this code block?

"Python console" -- it provides specialised syntax highlighting for the interactive REPL :)